component/main.jsonnet

```
@@ -1,10 +1,110 @@
 // main template for csi-driver-smb
+local com = import 'lib/commodore.libjsonnet';
+local espejo = import 'lib/espejo.libsonnet';
 local kap = import 'lib/kapitan.libjsonnet';
 local kube = import 'lib/kube.libjsonnet';
 local inv = kap.inventory();
 // The hiera parameters for the component
 local params = inv.parameters.csi_driver_smb;
 
-// Define outputs below
+local pvs = [
+  kube.PersistentVolume(pv.name) {
+    metadata+: {
+      namespace: pv.namespace,
+    },
+    spec+: {
+      capacity: {
+        storage: pv.capacity,
+      },
+      accessModes: com.getValueOrDefault(pv, 'accessModes', [ 'ReadWriteMany' ],),
+      persistenVolumeReclaimPolicy: com.getValueOrDefault(pv, 'reclaimPolicy', null),
+      mountOptions: com.getValueOrDefault(pv, 'mountOptions', [ 'dir_mode=0777', 'file_mode=0777', 'vers=3.0' ]),
+      csi: {
+        driver: 'smb.csi.k8s.io',
+        readOnly: com.getValueOrDefault(pv, 'readOnly', null),
+        volumeAttributes: {
+          source: '//%s/%s' % [ pv.share_host, pv.share_name ],
+        },
+        volumeHandle: 'pv-%s-%s' % [ pv.namespace, pv.name ],
+        nodeStageSecretRef: {
+          name: '%s-credentials' % [ pv.name ],
+          namespace: pv.namespace,
+        },
+      },
+      storageClassName: 'smb-%s' % [ pv.namespace ],
+    },
+  }
+  for pv in params.pvs
+];
+
+local pvcs = [
+  kube.PersistentVolumeClaim(pv.name) {
+    metadata+: {
+      namespace: pv.namespace,
+    },
+    spec+: {
+      accessModes: com.getValueOrDefault(pv, 'accessModes', [ 'ReadWriteMany' ],),
+      resources: {
+        requests: {
+          storage: pv.capacity,
+        },
+      },
+      storageClassName: 'smb-%s' % [ pv.namespace ],
+      volumeMode: 'Filesystem',
+      volumeName: pv.name,
+    },
+  }
+  for pv in params.pvs
+];
+local pvSecrets = [
+  kube.Secret(pv.name + '-credentials') {
+    metadata+: {
+      namespace: pv.namespace,
+    },
+    // need to use stringData here for secret reveal to work
+    // use keys which the csi-driver expects for the credentials secret
+    stringData+: {
+      username: pv.username_ref,
+      password: pv.password_ref,
+    },
+  }
+  for pv in params.pvs
+];
+
+local syncConfigs = [
+  espejo.syncConfig('restrict-smb-' + pv.namespace) {
+    spec: {
+      forceRecreate: true,
+      namespaceSelector: {
+        matchLabels: {
+          name: pv.namespace,
+        },
+      },
+      syncItems: [ {
+        apiVersion: 'v1',
+        kind: 'ResourceQuota',
+        metadata: {
+          name: 'restrict-smb-%s' % [ pv.namespace ],
```

I recommend to add the following labels:
```
local commonItemLabels = {
  'app.kubernetes.io/managed-by': 'espejo',
  'app.kubernetes.io/part-of': 'syn',
  'app.kubernetes.io/component': 'csi-driver-smb',
};
```
```suggestion
          name: 'restrict-smb-%s' % [ pv.namespace ],
          labels: commonItemLabels,
```